### PR TITLE
openneuro individual file download

### PR DESCRIPTION
**Pull request by MartinPerez**
to download partial openfmri datasets, raw and derivatives per subject alongside FSL statistical results for QA.

### CODE REVIEW

**Comment by MartinPerez** [comment]
Ready for review

File: nistats/datasets.py [hunk 1]
```diff
@@ -13,6 +13,9 @@
 from nilearn.datasets.utils import (
     _get_dataset_dir, _fetch_files, _fetch_file, _uncompress_file)
 
+import boto3
```
Comment (dohmatob) [inline nistats/datasets.py]: Is this dep documented / added to readme file, etc. ?

File: nistats/datasets.py [hunk 2]
```diff
@@ -93,46 +100,83 @@ def fetch_openfmri_dataset(dataset_name='ds000001', dataset_revision=None,
...
+            urls = json.load(json_file)
+
+    # filter urls
+    for exclusion in exclusion_filters:
+        urls = [url for url in urls if not re.match(exclusion, url)]
+    for inclusion in inclusion_filters:
+        urls = [url for url in urls if re.match(inclusion, url)]
+
+    # subject selection filter
+    subject_regex = 'sub-[a-z|A-Z|0-9]*[_./]'
```
Comment (dohmatob) [inline nistats/datasets.py]: This should be thoroughly tested.

File: nistats/datasets.py [hunk 3]
```diff
@@ -93,46 +100,83 @@ def fetch_openfmri_dataset(dataset_name='ds000001', dataset_revision=None,
...
+        def get_url(endpoint_url, bucket_name, file_key):
+            return '{}/{}/{}'.format(endpoint_url, bucket_name, file_key)
+
+        resource = boto3.resource('s3')
+        resource.meta.client.meta.events.register('choose-signer.s3.*',
+                                                  disable_signing)
+        bucket = resource.Bucket('openneuro')
+
+        for obj in bucket.objects.filter(Prefix=data_prefix):
+            if obj.key[-1] != '/':
```
Comment (dohmatob) [inline nistats/datasets.py]: Please comment what you're trying to do here.

**Comment by dohmatob** [comment]
You should document the `boto3` dependency and add it to the ci chain.

File: nistats/datasets.py (hunk 2 above)
Comment (MartinPerez) [reply nistats/datasets.py]: yeah and the inclusion and exclusion filters too alongside.

File: nistats/datasets.py (hunk 1 above)
Comment (MartinPerez) [reply nistats/datasets.py]: not yet, I realize not even added to travis, have to fix it

File: nistats/datasets.py [hunk 4]
```diff
@@ -55,33 +58,37 @@ def fetch_bids_langloc_dataset(data_dir=None, verbose=1):
...
+
+    n_subjects: int, optional
+        number of subjects to download from the dataset.
+
+    inclusion_filters: list of str, optional
+        List of regex strings that will be used to filter the url list.
+        If the regex matches the url it is retained for download.
+        Multiple filters work on top of each other.
+        Like an "and" logical operator, creating a more restrictive query.
+        Inclusion and exclusion filters apply together.
```
Comment (bthirion) [inline nistats/datasets.py]: This is typically a case where the API becomes daunting to users without a very explicit example. Please provide one.

File: nistats/datasets.py [hunk 5]
```diff
@@ -55,33 +58,37 @@ def fetch_bids_langloc_dataset(data_dir=None, verbose=1):
...
+
+    exclusion_filters: list of str, optional
+        List of regex strings that will be used to filter the url list.
+        If the regex matches the url it is discarded for download.
+        Multiple filters work on top of each other.
+        Like an "and" logical operator, creating a more restrictive query.
+        Inclusion and exclusion filters apply together.
+
```
Comment (bthirion) [inline nistats/datasets.py]: same thing here.

File: nistats/tests/test_datasets.py [hunk 6]
```diff
@@ -43,23 +44,19 @@ def test_fetch_bids_langloc_dataset():
...
+    data_prefix = '{}/{}/uncompressed'.format(
+        dataset_version.split('_')[0], dataset_version)
+    data_dir = _get_dataset_dir(data_prefix, data_dir=tst.tmpdir,
+                                verbose=1)
+    url_file = os.path.join(data_dir, 'urls.json')
+    json.dump([data_prefix + '/sub-xxx.html'], open(url_file, 'w'))
+
+    datadir, dl_files = datasets.fetch_openneuro_dataset(
+        tst.tmpdir, dataset_version)
+
```
Comment (bthirion) [inline nistats/tests/test_datasets.py]: As required by @dohmatob the test should cover many more options.

**Comment by bthirion** [comment]
Alos, I would make boto only an optional dependency, as most of nistats does not need it.

**Comment by MartinPerez** [comment]
@bthirion I fixed all issues. Expanded the filters doc with a small example. Ready for confirmation.

File: nistats/datasets.py [hunk 7]
```diff
@@ -55,33 +58,41 @@ def fetch_bids_langloc_dataset(data_dir=None, verbose=1):
...
+    inclusion_filters: list of str, optional
+        List of regex strings that will be used to filter the url list.
+        If the regex matches the url it is retained for download.
+        Multiple filters work on top of each other.
+        Like an "and" logical operator, creating a more restrictive query.
+        Inclusion and exclusion filters apply together.
+        For example the regex string '.*task-rest.*' would keep only urls
+        that contain the 'task-rest' string.
+
+    exclusion_filters: list of str, optional
```
Comment (bthirion) [inline nistats/datasets.py]: Note that inclusion and exclusion filters play the same role.

**Comment by bthirion** [comment]
LGTM

**Comment by MartinPerez** [comment]
@bthirion ready for review

**Pull request merged**